### PR TITLE
Add message definition read API

### PR DESCRIPTION
**Pull request by james-rms**
Adds an API to read message definitions out of a bag.

### CODE REVIEW

File: rosbag2_cpp/include/rosbag2_cpp/reader.hpp [hunk 1]
```diff
@@ -169,6 +169,14 @@ class ROSBAG2_CPP_PUBLIC Reader
    */
   std::vector<rosbag2_storage::TopicMetadata> get_all_topics_and_types() const;
 
+  /**
+   * Ask bagfile for all message definitions that were recorded.
+   *
+   * \param vector of message definitions to fill.
```
Comment (emersonknapp) [inline rosbag2_cpp/include/rosbag2_cpp/reader.hpp]:
```suggestion
   * \param[out] vector of message definitions to fill. Existing data will be overwritten.
```

**Review by emersonknapp — APPROVED** [review]

**Comment by james-rms** [comment]
Gist: https://gist.githubusercontent.com/james-rms/54be01ed14f3387d04752b1b5deb5350/raw/9315f5f965e18c5572aa4cef4862154321832663/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_compression rosbag2_cpp rosbag2_transport rosbag2_py rosbag2_storage_mcap rosbag2_storage_sqlite3
TEST args:  --packages-above rosbag2_compression rosbag2_cpp rosbag2_transport rosbag2_py rosbag2_storage_mcap rosbag2_storage_sqlite3
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/11898
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18555/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13099/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19290/)

File: rosbag2_py/test/test_sequential_reader.py [hunk 2]
```diff
@@ -89,6 +89,25 @@ def test_sequential_reader_seek(storage_id):
...
+        assert parameter_event_msg.topic_type == 'rcl_interfaces/msg/ParameterEvent'
+        assert parameter_event_msg.encoding == 'ros2msg'
+        assert 'Parameter[] new_parameters' in parameter_event_msg.encoded_message_definition
+
+        assert string_msg.topic_type == 'std_msgs/msg/String'
+        assert string_msg.encoding == 'ros2msg'
+        assert 'string data' in string_msg.encoded_message_definition
+    else:
+        assert len(message_definitions) == 0
+
```
Comment (MichaelOrlov) [inline rosbag2_py/test/test_sequential_reader.py]: It looks like unrelated to the `test_sequential_reader_seek`. It will be better to create separate test case for testing message definitions.

File: rosbag2_storage_mcap/src/mcap_storage.cpp [hunk 3]
```diff
@@ -641,6 +643,27 @@ std::vector<rosbag2_storage::TopicMetadata> MCAPStorage::get_all_topics_and_type
...
+void MCAPStorage::get_all_message_definitions(
+  std::vector<rosbag2_storage::MessageDefinition> & definitions)
+{
+  ensure_summary_read();
+  auto schema_map = mcap_reader_->schemas();
+  definitions.clear();
+  definitions.reserve(schema_map.size());
+  for (const auto & [id, schema_ptr] : schema_map) {
+    std::string encoded_message_definition = "";
+    if (schema_ptr->data.size() != 0) {
```
Comment (MichaelOrlov) [inline rosbag2_storage_mcap/src/mcap_storage.cpp]: Small nitpicks from Clang-Tidy
```suggestion
    std::string encoded_message_definition;
    if (!schema_ptr->data.empty()) {
```

File: rosbag2_storage_mcap/test/rosbag2_storage_mcap/test_mcap_storage.cpp [hunk 4]
```diff
@@ -36,6 +36,12 @@ bool operator==(const TopicInformation & lhs, const TopicInformation & rhs)
 {
   return lhs.topic_metadata == rhs.topic_metadata && lhs.message_count == rhs.message_count;
 }
+
+bool operator==(const MessageDefinition & a, const MessageDefinition & b)
+{
+  return a.topic_type == b.topic_type && a.encoding == b.encoding &&
+         a.encoded_message_definition == b.encoded_message_definition;
+}
```
Comment (MichaelOrlov) [inline rosbag2_storage_mcap/test/rosbag2_storage_mcap/test_mcap_storage.cpp]: @james-rms Could you please move this comparison operator inside `MessageDefinition` declaration?
I have a suspicious that it will be required from another places as well. At least from test for SQlite3 plugin.

**Review by MichaelOrlov — CHANGES_REQUESTED** [review]
@james-rms Overall looks good among a few nitpicks and comments about minor improvements mostly related to the tests.

**Review by MichaelOrlov — APPROVED** [review]
LGTM.
Thanks!

**Comment by james-rms** [comment]
Gist: https://gist.githubusercontent.com/james-rms/22da1a40ef915f159a50dd74762390f5/raw/9315f5f965e18c5572aa4cef4862154321832663/ros2.repos
BUILD args:  --packages-above-and-dependencies rosbag2_compression rosbag2_cpp rosbag2_transport rosbag2_py rosbag2_storage_mcap rosbag2_storage_sqlite3
TEST args:  --packages-above rosbag2_compression rosbag2_cpp rosbag2_transport rosbag2_py rosbag2_storage_mcap rosbag2_storage_sqlite3
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/11900
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/18557/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/13101/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/19292/)

**Comment by ros-discourse** [comment]
This pull request has been mentioned on **ROS Discourse**. There might be relevant details there:

https://discourse.ros.org/t/ros-2-tsc-meeting-minutes-2023-05-18/31587/1
